File: Data Structures & Algorithms/longest-increasing-subsequence/submission-3.py
class Solution:
    def lengthOfLIS(self, nums: List[int]) -> int:
        memo = {}
        res = 0

        def dfs(i, curr) -> int:
            if i >= len(nums):
                return 0

            if (i,curr) in memo:
                return memo[(i,curr)]

            skip = dfs(i+1, curr)
            take = 0

            if nums[i] > curr:
                take = 1 + (dfs(i+1, nums[i]))

            memo[(i,curr)] = max(skip, take)
            return memo[(i,curr)]
        
        for i in range(len(nums)):
            res = max(res, dfs(i,float("-inf")))

        return res

File: Data Structures & Algorithms/integer-break/submission-3.java
class Solution {
    private int[] memo;
    public int integerBreak(int n) {
        memo = new int[n + 1];
        return dfs(n);
    }

    private int dfs(int n) {
        if (memo[n] != 0) {
            return memo[n];
        }
        int max = 1;

        for (int i = 1; i < n; i++) {
            max = Math.max(max, i * Math.max(n - i, dfs(n - i)));
        }

        return memo[n] = max;
    }
}